File: leetcode/src/main/java/divideandconquer/_89_majority_element_169/Solution2.java
package divideandconquer._89_majority_element_169;

// TODO : 분할정복
public class Solution2 {
    public static int majorityElement(int left, int right, int[] nums) {
        // 최소 분할 단위 리턴
        if (left == right)
            return nums[left];

        // 중앙 위치 인덱스 계산
        int mid = left + (right - left) / 2;
        int a = majorityElement(left, mid, nums);
        int b = majorityElement(mid + 1, right, nums);

        int countA = 0;
        // 해당 구간의 a 개수 계산
        for (int i = left; i <= right; i++) {
            if (nums[i] == a)
                countA++;
        }
        // a가 과반수를 넘으면 a 리턴
        return countA > (right - left + 1) / 2 ? a : b;
    }

    public static int majorityElement(int[] nums) {
        return majorityElement(0, nums.length - 1, nums);
    }

    public static void main(String[] args) {
        majorityElement(new int[]{2, 2, 1, 1, 1, 2, 2});
    }
 // 3ms
}

File: leetcode/src/main/java/nonlinear/tree/bst/_55_convert_sorted_array_to_binary_search_tree_108/Solution2.kt
package nonlinear.tree.bst._55_convert_sorted_array_to_binary_search_tree_108

class Solution2 {
    fun sortedArrayToBST(nums: IntArray): Main.TreeNode? {
        fun construct(lo: Int, hi: Int): Main.TreeNode? {
            // 예외 처리
            if (lo > hi) return null
            // 인덱스의 중앙값 계산, 소수점은 내림
            val mid = lo + (hi - lo) / 2
            // 배열의 중앙값으로 노드 생성
            val node = Main.TreeNode(nums[mid])
            // 왼쪽 자식 노드는 배열 앞부분의 중앙값
            node.left = construct(lo, mid - 1)
            // 오른쪽 자식 노드는 배열 뒷부분의 중앙값
            node.right = construct(mid + 1, hi)
            return node
        }

        // 예외 처리
        return if (nums.isEmpty()) null
        else
        // 배열 전체를 넘겨서 트리 생성 시작
            construct(0, nums.size - 1)
    }
}